braindecode/models/usleep.py

```diff
@@ -0,0 +1,272 @@
+                    nn.ELU(),
+                    nn.BatchNorm1d(num_features=out_channels),
+                )
+
+    def forward(self, x, residual):
+        x = self.block_preskip(x)
+        if self.with_skip_connection:
+            x, residual = _crop_tensors_to_match(x, residual, axis=-1)  # in case of mismatch
+            x = torch.cat([x, residual], axis=1)  # (B, 2 * C, T)
+        x = self.block_postskip(x)
+        return x
+
+
+class USleep(nn.Module):
+    """Sleep staging architecture from Perslev et al 2021.
+
+    U-Net (autoencoder with skip connections) feature-extractor for sleep
+    staging described in [1]_.
+
+    For the encoder ('down'):
+        -- the temporal dimension shrinks (via maxpooling in the time-domain)
+        -- the spatial dimension expands (via more conv1d filters in the
+           time-domain)
+    For the decoder ('up'):
+        -- the temporal dimension expands (via upsampling in the time-domain)
+        -- the spatial dimension shrinks (via fewer conv1d filters in the
+           time-domain)
+    Both do so at exponential rates.
+
+    Parameters
+    ----------
+    in_chans : int
+        Number of EEG or EOG channels. Set to 2 in [1]_ (1 EEG, 1 EOG).
+    sfreq : float
+        EEG sampling frequency. Set to 128 in [1]_.
+    complexity_factor : float
+        Multiplicative factor for number of channels at each layer of the U-Net.
+        Set to sqrt(2) in [1]_.
+    with_skip_connection : bool
+        If True, use skip connections in decoder blocks.
+    n_classes : int
+        Number of classes. Set to 5.
+    input_size_s : float
+        Size of the input, in seconds. Set to 30.
+    apply_softmax : bool
+        If True, apply softmax on output (e.g. when using nn.NLLLoss). Use
+        False if using nn.CrossEntropyLoss.
+
+    References
+    ----------
+    .. [1] Perslev M, Darkner S, Kempfner L, Nikolic M, Jennum PJ, Igel C.
+           U-Sleep: resilient high-frequency sleep staging. npj Digit. Med. 4, 72 (2021).
+           https://github.com/perslev/U-Time/blob/master/utime/models/usleep.py
+    """
+    def __init__(self,
+                 in_chans=2,
+                 sfreq=100,
+                 complexity_factor=np.sqrt(2),
+                 with_skip_connection=True,
+                 n_classes=5,
+                 input_size_s=30,
+                 apply_softmax=False
+                 ):
+        super().__init__()
+
+        self.in_chans = in_chans
+
+        # Harcoded (otherwise dims can break)
+        depth = 10
+        time_conv_size = 9  # 0.09s at sfreq = 100 Hz
+        max_pool_size = 2   # 0.02s at sfreq = 100 Hz
+        n_time_filters = 5
+
+        # Convert between units: seconds to time-points (at sfreq)
+        input_size = np.ceil(input_size_s * sfreq).astype(int)
+
+        # Define geometric sequence of channels
+        channels = (
+            n_time_filters * complexity_factor * np.sqrt(2) ** np.arange(0, depth + 1)
```

Comment:
Should there be both `complexity_factor` and `np.sqrt(2)` here?

Reply:
Yes, we just make the same things as USleep

Reply:
I'm confused by the description in the paper:
![image](https://user-images.githubusercontent.com/6500858/124154202-45a52380-da63-11eb-9c6c-c3f624b77f60.png)

Each layer has srqt(2) times more convolutional channels than the previous one. For the encoder at least, there's only one conv layer per block. How can the number of degrees of freedom **double** from one block to the next?

Reply:
Why do you want the degree of freedom double? it is just multiply by sqrt(2)
Maybe i don't understand your point

Reply:
The doubling is just what they say in the paper. It doesn't seem to fit with their equation so that's what I was wondering about. You've looked at their keras implementation - maybe it's clearer in there?

With the current code though, you end up with:
```
np.ceil(5 * np.sqrt(2) * np.sqrt(2) ** np.arange(0, 10 + 1))
Out[5]: array([  8.,  11.,  15.,  21.,  29.,  41.,  57.,  81., 114., 161., 227.])
```
That doesn't look right to me. I think it should be `np.floor` instead of `np.ceil`, and the hardcoded `np.sqrt(2)` should be removed since that's what `complexity_factor` does.

That would give: `np.floor(n_time_filters * complexity_factor ** np.arange(0, depth + 1))`
```
np.floor(5 * np.sqrt(2) ** np.arange(0, 10 + 1))
Out[6]: array([  5.,   7.,  10.,  14.,  20.,  28.,  40.,  56.,  80., 113., 160.])
```

Reply:
Ok, it's because Omar doesn't like the way USleep is code and he tries somethings different. So it is not exactly the code in keras, i don't know what is the best

Reply:
????? !!!! you mean this architecture is not the published architecture ! and this is done silently as we say in the doc it's USleep???